### PR TITLE
Refactor and improvements

### PR DESCRIPTION
**Pull request by konradmalik**
Hey!

This is a good plugin, I've been using it for quite some time with success.

This PR is a refactor proposition which (in my opinion at least) simplifies some code and moves it into more appropriate places.

Also notable changes:
- support for stopping the server when no longer needed (via on_exit client callback)
- scope CSTarget to buffer and remove after use

Please let me know what you think! Also, don't hesitate to mention if it does not align with your vision here.

### CODE REVIEW

File: lua/roslyn/init.lua [hunk 1]
```diff
@@ -271,14 +211,22 @@ function M.setup(config)
...
             vim.notify_once(
                 "Multiple sln files found. You can use `CSTarget` to select target for buffer",
                 vim.log.levels.INFO
             )
+
+            vim.api.nvim_buf_create_user_command(opt.buf, "CSTarget", function()
+                vim.ui.select(all_sln_files, { prompt = "Select target solution: " }, function(sln_file)
+                    known_solutions[sln_directory] = sln_file
+                    wrap_roslyn(sln_file, roslyn_config)
+                    vim.api.nvim_buf_del_user_command(opt.buf, "CSTarget")
```
Comment (seblj) [inline lua/roslyn/init.lua:227]: I am a bit sceptical to delete the command after using it. What if I select the wrong target for the buffer and then I want to change it again?

Comment (seblj) [reply lua/roslyn/init.lua:227]: The change to make it scoped to the buffer makes sense though!

**Review by seblj — CHANGES_REQUESTED** [review]
Thanks! I think most of this makes sense, but there is some things that needs to be changed before I can merge😅

File: lua/roslyn/init.lua [hunk 2]
```diff
@@ -12,21 +14,59 @@ local function get_mason_installation()
...
+---@param config? vim.lsp.ClientConfig
 ---@param filewatching boolean
 local function lsp_start(pipe, target, config, filewatching)
+    local sln_directory = vim.fs.dirname(target)
+    config = config or {}
     config.name = "roslyn"
     config.cmd = vim.lsp.rpc.connect(pipe)
-    config.root_dir = vim.fs.dirname(target)
+    config.root_dir = sln_directory
+    config.capabilities = get_default_capabilities(filewatching)
```
[the remark this answers is not in the document]
Comment (konradmalik) [reply lua/roslyn/init.lua]: ahh yes, totally missed that. See [7cee7fd](https://github.com/seblj/roslyn.nvim/pull/25/commits/7cee7fdb90defc7c84df3f3acaff95711d1c70b2) where I fixed it

File: lua/roslyn/server.lua [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+local M = {}
+
+---@param cmd string[]
+---@param with_pipe_name fun(pipe_name: string): nil A function to execute after server start and pipe_name is known
+function M.start_server(cmd, with_pipe_name)
+    if _pipe_name then
+        with_pipe_name(_pipe_name)
+        return
+    end
+    if not _server_object then
```
[the remark this answers is not in the document]
Comment (konradmalik) [reply lua/roslyn/server.lua]: fixed in https://github.com/seblj/roslyn.nvim/pull/25/commits/0330e8ae6a2925ef96be6256dbb6724bc6d36e6d

File: lua/roslyn/server.lua [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        vim.notify("starting new roslyn process", vim.log.levels.INFO)
```
[the remark this answers is not in the document]
Comment (konradmalik) [reply lua/roslyn/server.lua]: fixed in https://github.com/seblj/roslyn.nvim/pull/25/commits/0330e8ae6a2925ef96be6256dbb6724bc6d36e6d

File: lua/roslyn/server.lua [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+function M.stop_server()
+    if not _server_object then
+        return
+    end
+
+    _server_object:kill(9)
+    _pipe_name = nil
+    _server_object = nil
+    vim.schedule(function()
+        vim.notify("stopping roslyn process", vim.log.levels.INFO)
```
[the remark this answers is not in the document]
Comment (konradmalik) [reply lua/roslyn/server.lua]: fixed in https://github.com/seblj/roslyn.nvim/pull/25/commits/0330e8ae6a2925ef96be6256dbb6724bc6d36e6d

File: lua/roslyn/init.lua (hunk 1 above)
Comment (konradmalik) [reply lua/roslyn/init.lua]: yeah, this makes sense, deleting is not the best idea. Fixed in https://github.com/seblj/roslyn.nvim/pull/25/commits/d4cdda1d9ba2fd5c60c2662e94f9d062bc6c1bf2

File: lua/roslyn/init.lua [hunk 6]
```diff
@@ -12,21 +14,58 @@ local function get_mason_installation()
...
+end
 
 ---@type table<string, string>
 ---Key is solution directory, and value is sln target
 local known_solutions = {}
 
 ---@param pipe string
 ---@param target string
----@param config vim.lsp.ClientConfig
+---@param config? vim.lsp.ClientConfig
```
Comment (seblj) [inline lua/roslyn/init.lua:61]: This don't need to be optional anymore I think

File: lua/roslyn/init.lua [hunk 7]
```diff
@@ -209,16 +172,14 @@ function M.setup(config)
...
         exe = nil,
-        -- I don't know how to make this a partial type.
         ---@diagnostic disable-next-line: missing-fields
         config = {
+            -- setting default capabilities early allows users to provide their own
             capabilities = get_default_capabilities(config),
         },
     }
 
-    ---@type InternalRoslynNvimConfig
```
Comment (seblj) [inline lua/roslyn/init.lua:219]: Maybe add this type back? Unless the language server is infering this. On mobile now so don't have time to check now

**Review by seblj — COMMENTED** [review]
Okay this looks pretty good now! Some small nits about typing. Can't see something else from my phone now, so I will test this (hopefully tomorrow), and if no immediate issues, I think it is good😄

File: lua/roslyn/init.lua (hunk 6 above)
Comment (konradmalik) [reply lua/roslyn/init.lua]: yeah, sorry, rebased and amended the previous commit 😅 see https://github.com/seblj/roslyn.nvim/pull/25/commits/69a681b0897a5cad400cd90a9d101c6d15135b96

File: lua/roslyn/init.lua [hunk 8]
```diff
@@ -189,18 +142,28 @@ local function get_cmd(exe)
     end
 end
 
-local M = {}
-
 ---@class InternalRoslynNvimConfig
----@field filewatching boolean
----@field exe? string
+---@field filewatching? boolean
```
[the remark this answers is not in the document]
Comment (konradmalik) [reply lua/roslyn/init.lua]: see b37c4d9495de2b5db5a0df9aea797a8eabddc436

File: lua/roslyn/init.lua (hunk 7 above)
Comment (konradmalik) [reply lua/roslyn/init.lua:219]: yeah, removed because lua ls is inferring properly. Can add though if you really want ;)

Comment (seblj) [reply lua/roslyn/init.lua:219]: No it's okay then😄

**Pull request merged**